docs/git/index.md

```diff
@@ -0,0 +1,220 @@
+---
+id: index
+title: Introduction
+sidebar_label: Introduction
+---
+
+Most software engineers use a tool called Git to organise their code, and collaborate with other people.
+
+Git can be used for many things, and we will see lots of them throughout the course.
+
+Each week, you're going to use Git to get a copy of your homework exercises, to submit your solutions, and to get feedback on your solutions. Today, you're going to learn how to do these things.
+
+## Learning Objectives
+
+By the end of this class, you should be able to:
+
+* Get a copy of your homework questions onto your computer.
+  * (To _clone_ a repository).
+* Save your answers, and send them to our volunteers.
+  * (To _commit_ and _push_ changes).
+* See and respond to feedback on your homework.
+  * (To _fork_ a repository, use Github's _pull request_ workflows, and _push_ further changes).
+* Explore how a file has changed over time.
+  * (To use Git's _history log_).
+
+---
+
+## Prerequisites
+
+1. A Github account. If you don't have one already, go to https://github.com/ and sign up for one.
+2. Installed Github Desktop. If you don't have it yet, go to https://desktop.github.com/ and follow its instructions to install it.
+3. Sign in to Github Desktop with your Github account. Run Github Desktop (TODO: Instructions on how to find it), click "Sign in to GitHub.com", and go through the steps to sign in.
+4. Installed VS Code.
+
+## What problem does Git solve?
+
+Git is used to solve many problems, but today we're going to focus on how it:
+* Helps us to share information.
+* Enables people to make their own changes to that information, and share it back.
+* Allows us to track how information has changed over time.
```

Comment:
From @40thieves:

One of my observations from CYF is that students are "scared" to make changes to their code once something is working. This leads to large amounts of commenting out, or hesitancy to try new things out.

Therefore I wonder if there's benefit in explaining how VCS gives us safety: "experiments" can be done and quickly discard back to a known "safe point". We could look at ways in which versioning helps with this, providing further context on why Git is useful.

In Github Desktop, there are are couple of features that look potentially useful:

1. When there are uncommitted changes, there is an "Discard changes" option when right-clicking the changed file. It appears to do something like `git checkout -- PATH/TO/FILE`. It could be used to teach: "commit often, make the change you're not confident about, and if it doesn't work discard it".
2. Once you've committed, an "Undo" button is shown under the commit message box. It appears to do something like `git reset HEAD^`. It could be used to teach "commit, check the diff, and undo if it doesn't look right"

I'm a bit on the fence about whether we should actually teach this in Github Desktop. Firstly, it feels a bit like dangerous territory, as students could lose significant amount work if they are not committing regularly. Secondly, we are potentially short on time, and none of my ideas directly help with submitting homework so we could be better served by more practice.

Reply:
I think that introducing "discard uncommitted changes" (the right-click `git checkout -- file`) would be really useful, as a way of establishing checkpoints, and I think that "see each change you've done in history so you can manually undo it by copying/pasting" is a great technique for us to teach.

> Once you've committed, an "Undo" button is shown under the commit message box. It appears to do something like git reset HEAD^. It could be used to teach "commit, check the diff, and undo if it doesn't look right"

Interesting... It looks like it goes away (and doesn't come back), but it adds a revert-like commit on top, rather than messing with history at all. This one-off button could be handy to demonstrate, but I don't think we should make a particularly big point of it.

I'm wary of teaching anything around revert in our first session - I think it's both an overwhelming amount of information, and potentially risky... Revert (beyond the automatic undo button) immediately gets us at risk of needing to teach how to resolve merge conflicts, and that is definitely not first-git-experience content...

Reply:
That sounds reasonable to me - as mentioned in my original comment, I am also hesitant to get into revert.

I like your idea of "checkpoints" 👍